### ISSUE TITLE
Union types: discriminate between table types and between enums | strings

### ISSUE THREAD

**Issue by pdesaulniers** [teal-language/tl#108]
As mentioned in the tutorial, the `is` operator can only discriminate across primitive types and at most one table type. The same limitation exists between enums and strings.

It would be nice if there wasn't such limitation.

**Comment by hishamhm** [comment]
> discriminate across table types

This will require metatable-equipped records (or objects). For reference, see related discussion in #97.

> between enums | strings

This will require reifying the enum in the generated code for runtime lookup. Not done yet, but doable. This one seems pretty low-priority, though.

**Comment by Fang-** [comment]
> or some sort of tagged unions

Considering they're mentioned here, I'll refrain from opening a separate issue for this, but: supporting tagged unions seems very important to me. Coming from a language that supports them, they're rightfully used all over the place, and can make up for unimplemented complex union cases like the mentioned `string | MyEnum`.  
#124 seems to be pushing in this direction already, could be worth pursuing that further.

(Regardless, excited about this project! Thanks @hishamhm and all contributors involved for the great work so far. (: )

**Comment by makspll** [comment]
+1, I am embedding Teal in Rust, and the limitation of not being able to use multiple UserData types in an union is a massive problem for me, I have to basically resort to using 'Any' everywhere which then defeats the purpose of using teal in the first place :p.

**Comment by lenscas** [comment]
I think a decent compromise would be to allow `table | otherTable` (or `userdata | otherUserData`) but do **not** allow `is` to disambiguate between them.

This should allow people to create their own functions to disambiguate between the variants (which sadly would require `as`) and would also make documenting api's that are not written in teal easier (Like libraries written in lua but also apis written in C and Rust which might use a library to generate the documentation)

**Comment by hishamhm** [comment]
Half of this is implemented: one can now equip record types with an `is` handler that allows discriminating between them.

Discriminating between enums and strings is still to-do!

**Comment by farism** [comment]
> Half of this is implemented: one can now equip record types with an `is` handler that allows discriminating between them.
> 
> Discriminating between enums and strings is still to-do!

Are there any docs on how to use `is` in records to allow discriminating?

**Comment by mbartlett21** [comment]
> > Half of this is implemented: one can now equip record types with an `is` handler that allows discriminating between them.
> > Discriminating between enums and strings is still to-do!
> 
> Are there any docs on how to use `is` in records to allow discriminating?

https://teal-language.org/book/current_limitations_on_union_types.html

tl;dr:

Use a `where` clause in your record or interface definition with an expression that uses `self` once.

E.g:

```tl
local record A
   where self.ty == 'A'
   ty: string
   other_field: number
end

local record B
   where self.ty == 'B'
   ty: string
end

local a: A | B = { ty = 'A', other_field = 5 }

assert(a is A)
```
